cupyx/scipy/ndimage/measurements.py

```diff
@@ -200,3 +203,282 @@ def _kernel_finalize():
+            (-1,) + (1,) * input.ndim)
+        sum_axes = tuple(range(1, 1 + input.ndim))
+        out[i:i + batch_size] = cupy.where(matched, input, 0).sum(
+            axis=sum_axes)
+    return out
+
+
+_ndimage_mean_kernel = cupy.ElementwiseKernel(
+    'T input, R labels, raw X index, int32 size',
+    'raw Y out, raw int32 count',
+    """
+    for (ptrdiff_t j = 0; j < size; j++) {
+      if (labels == index[j]) {
+        atomicAdd(&out[j], input);
+        atomicAdd(&count[j], 1);
+        break;
+      }
+    }
+    """)
+
+
+def _ndimage_mean_kernel_2(input, labels, index, batch_size=4,
+                           return_count=False):
+    out = cupy.empty_like(index, dtype=input.dtype)
+    count = cupy.empty_like(index, dtype=cupy.int32)
+    for i in range(0, index.size, batch_size):
+        matched = labels == index[i:i + batch_size].reshape(
+            (-1,) + (1,) * input.ndim)
+        mean_axes = tuple(range(1, 1 + input.ndim))
+        count[i:i + batch_size] = matched.sum(axis=mean_axes)
+        out[i:i + batch_size] = cupy.where(matched, input, 0).sum(
+            axis=mean_axes)
+    if return_count:
+        return out/count, count
+    return out/count
+
+
+def _mean_driver(input, labels, index, return_count=False, use_kern=False):
+    # The following parameters for sum where determined using a Tesla P100.
+    if use_kern:
+        return _ndimage_mean_kernel_2(input, labels, index,
+                                      return_count=return_count)
+
+    out = cupy.zeros_like(index, input.dtype)
+    count = cupy.zeros_like(index, dtype=cupy.int32)
+    sum, count = _ndimage_mean_kernel(input,
+                                      labels, index, index.size, out, count)
+    if return_count:
+        return sum/count, count
+    return sum/count
+
+
+def variance(input, labels=None, index=None):
+    """Calculate the variance of the values of an n-D image array, optionally
+    at specified sub-regions.
+
+    Args:
+        input (cupy.ndarray) :Nd-image data to process.
+        labels (cupy.ndarray or None): Labels defining sub-regions in `input`.
+            If not None, must be same shape as `input`.
+        index (cupy.ndarray or None): `labels` to include in output.  If None
+            (default), all values where `labels` is non-zero are used.
+
+    Returns:
+        variance (cupy.ndarray): Values of variance, for each sub-region if
+        `labels` and `index` are specified.
+
+
+    .. seealso:: :func:`scipy.ndimage.variance`
+    """
+    if not isinstance(input, cupy.ndarray):
+        raise TypeError('input must be cupy.ndarray')
+
+    # There is constraints on types because atomicAdd() in CUDA 7.5
+    # only supports int32, uint32, uint64, and float32.
+    if not issubclass(input.dtype.type,
+                      (cupy.int32, cupy.float16, cupy.float32, cupy.float64,
+                       cupy.uint32, cupy.uint64, cupy.ulonglong)):
+        raise TypeError(
```

Comment:
Some help in this would be appriciated as I could not come up with a implmentation of variance that used matching that was clean and didn't require me to completely re-implmenent a new version of _ndarray_var where I could set alpha.

Reply:
Like this for 1D?
```Python
mean_val, count = _mean_driver(input, labels, index, True)
cupy.where(labels == index[..., None], cupy.square(input - mean_val[..., None]), 0).sum(axis=(1,)) / count
```

Maybe we can use `ravel()` for inputs more than 1D.

Reply:
I dont think this can work, for e.g. the input can be of shape of 256, but the mean_val is going to be of shape index, and it makes no sense for them to be able to be broadcastable together or the subtraction to be logically correct.

Reply:
The ravel wasn't necessary as it was able to internally broadcast them properly.